### PR TITLE
Handle FCM push on update news feed

### PR DESCRIPTION
**Pull request by KeithYokoma**
## Overview (Required)
- Handle push notification using FCM triggered by news feed update.
- Visual may change because this comes from my idea!

## Changes
- Added FCM push handler service using Firebase SDK
- Abstraction for push handler logic
  - MessageClassifier gives corresponding push message handler to a "from" property in remote message
  - If RemoteMessage comes from news feed update, it returns NewPostProcessor
  - If no corresponding processor found for it, it falls back to DefaultProcessor

## Screenshot

device-2018-02-01-133914

### CODE REVIEW

File: app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushServiceModule.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package io.github.droidkaigi.confsched2018.service.push
+
+import dagger.Binds
+import dagger.Module
+import dagger.android.AndroidInjector
+
+@Module
+interface PushServiceModule : AndroidInjector<PushService> {
+    @Binds fun providesService(service: PushService): PushService
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushServiceModule.kt]: I think if you don't have PushService inject, PushServiceModule is not needed.
What do you think?

Comment (KeithYokoma) [reply app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushServiceModule.kt]: I didn't know that! I will remove it :)

**Comment by KeithYokoma** [comment]
Additional Note:
1. Push will not come until a feed is published.
2. DroidKaigi app can receive and show notification even if it is in the background.

**Comment by KeithYokoma** [comment]
After this is merged, I will work on deploying firebase functions to prod env.

**Comment by takahirom** [comment]
Probably we should call createNotificationChannel for Android O 👀 
You can use NotificationHelper. Please read it 🙏 
https://github.com/droidkaigi/conference-app-2018/blob/6ccbfbb5a865968d4895037d8dd44839cab9e39c/app/src/main/java/io/github/droidkaigi/confsched2018/presentation/common/notification/NotificationHelper.kt#L17-L17

**Comment by KeithYokoma** [comment]
Got it!

File: app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushService.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.github.droidkaigi.confsched2018.service.push
+
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import dagger.android.AndroidInjection
+import timber.log.Timber
+import javax.inject.Inject
+
+/**
+ * @author keishinyokomaku
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushService.kt]: Sorry please delete this 🙏

File: app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushServiceBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package io.github.droidkaigi.confsched2018.service.push
+
+import dagger.Module
+import dagger.android.ContributesAndroidInjector
+
+/**
+ * @author keishinyokomaku
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/service/push/PushServiceBuilder.kt]: Sorry please delete this 🙏

File: app/src/main/java/io/github/droidkaigi/confsched2018/service/push/processor/NewPostProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        val title = message.data[KEY_TITLE]
+        val content = message.data[KEY_CONTENT]
+        val type = message.data[KEY_TYPE]
+        val openPi = PendingIntent.getActivity(application,
+                0,
+                Intent(application, MainActivity::class.java)
+                        .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
+                // TODO open with announce fragment in front?
+                PendingIntent.FLAG_UPDATE_CURRENT)
+        val largeIcon = when (type) {
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/service/push/processor/NewPostProcessor.kt]: 👍

**Comment by takahirom** [comment]
Thanks! Almost LGTM 👍

**Comment by KeithYokoma** [comment]
Resolving conflicts now!

**Comment by takahirom** [comment]
🙏

**Comment by takahirom** [comment]
LGTM 👍 🔔

**Pull request merged**